### DISCUSSION TITLE
Suggestions for (additional/expected?) features

### DISCUSSION THREAD

**Discussion by arladmin** [Budibase/budibase#1133] [Ideas]
Hi. As i've recently discovered this tool, i decided to learn it by heading straight into the building experience, using the example of an existing app of mine.

A bit of my experience here:

Background: I am not a professional developer, just a savvy engineering/tech/business young professional with moderate programming experience. I started using a low/no code (SPA) platform, Knack.com, to develop an ERP of sorts for a medium sized company. I was able to build (while learning) and deploy quite a few complex (with extensive Access Controls, Workflows, Web/Mobile based, Several Users, Integrations) apps, like Order Management, Sales CRM, Production Management, Project Management, HRMS and more systems, from scratch. I started this in late 2018 and was able to fully deploy all of the systems, without any other tech personnel/external help, within 4-5 months.

Only major drawbacks: 
**speed**, because Knack only provides dedicated server resources for Enterprise plans, whereas we operate on a budget of a small startup, and the shared resources can lack quite a bit,
**cost increment** , because the prices increased in proportion to the records/objects created, and that rose up sharply. Much more than what one would experience from a normal database in a similar situation.

Now, i've been on the search for a new platform, that helps us with most of our requirements (high complexity, high speed, integrations, feature-rich; access controls, mobile friendly, minimal time for development/deployment, and of course, low budget), for the past several months. Have been, and still do, trying out mainly open-source(ish), low/no code platforms.

Glad to say that Budibase is one such platform that has given me high hopes (Saltcorn & Appsmith being others). I can deploy it to a private server close to us as well as maintain a low budget and also utilize latest features. Although, i suspect that it would still take the platform several months (even quarters) to grow to where i can fully implement what i have, in the past. (_Please correct me if i'm mistaken._)
 
While i'm not recommending/expecting Budibase to completely follow/duplicate other platforms, i would still like to mention some of the main features that i've been using earlier quite a bit and are missing here:
- Option to duplicate Tables
- Option to duplicate Columns
- Option to copy Columns to other tables
- Default columns (ID, Created By, Created At, Modified By, Modified At)
- Additional Column (data) types: actually, most of [**these**](https://learn.knack.com/article/13eszen598-field-types) are useful
- Column type: auto increment
- Calculated/Formula data types
- Conditional Visibility/Access rules for Pages/Screens
- Use/Refer data in a table from other (connected) tables

^ Almost all of these are necessary in each of the various apps that i've built/would like to build

I provide all this info only in the spirit of a healthy feedback and not negative criticism. If i'm mistaken anywhere, in regards to my perception about Budibase, kindly correct me. Otherwise, i would be glad if this strikes up a discussion and also results in several new concept/feature improvements/additions in Budibase.

Will add more points if/when i encounter along my journey with Budibase. Will appreciate any feedback to my message above as that will also help in planning out my expectations+strategy for the near-future, in the context of using Budibase .

Cheers!

**Comment by joebudi** [comment]
First, I would like to thank you for taking the time to provide us with such a helpful comment. I always tell users/stakeholders, feedback is worth more than gold, and in this case, you've provided us with a treasure chest.

In re to the features you have mentioned, here is an update:
- Option to duplicate Tables - this shouldn't be too difficult. We will try to find some time to squeeze this into one of the next two releases.

- Option to duplicate Columns - this shouldn't be too difficult. We will try to find some time to squeeze this into one of the next two releases.

- Option to copy Columns to other tables - I need to look into this a little more.

- Default columns (ID, Created By, Created At, Modified By, Modified At) - this is in the next release (0.8)

- Additional Column (data) types: actually, most of these are useful - we will continue to add types as we progress. Are there any specific you need now?

- Column type: auto increment - this is in the next release (0.8)

- Calculated/Formula data types - popular request and one we will tackle in the near future

- Conditional Visibility/Access rules for Pages/Screens - this is in the next release (0.8)

- Use/Refer data in a table from other (connected) tables - this is in the next release (0.8)



If you are free sometime, I would love to chat. My email is joe at budibase dot com. Send me an email, and if you are happy we can arrange a call and I can update you on our plans. We are building an unstoppable ecosystem, and we need incredible users like you to help provide feedback and drive the product forward.

Thanks again
Joe




along with other incredible features such as a new form-building experience.

**Reply by arladmin** [reply]
@joebudi, thanks a lot for your elaborate response. More over, it is really exciting for me to know that quite a few of the features are already upcoming. I did not expect this to be so fast! 

Regarding the additional column data types that are relevant for my use-cases:
(using Knack references again)

- https://learn.knack.com/article/13eszen598-field-types#rich-text
- https://learn.knack.com/article/13eszen598-field-types#currency
- https://learn.knack.com/article/nsrud4si6l-equations
- https://learn.knack.com/article/m9xm4ae3yq-formulas
- https://learn.knack.com/article/13eszen598-field-types#email
- https://learn.knack.com/article/13eszen598-field-types#address (Geocoded addresses)
- https://learn.knack.com/article/13eszen598-field-types#link

In addition to data types, there are [Validation rules](https://learn.knack.com/article/1isca4mrri-validation-rules) and [Conditional (update) rules](https://learn.knack.com/article/oaj8ajhwkc-conditional-rules), which can be applied to all the fields (columns) and these are really useful.

Formula and Equation data types are quite critical for most applications as they allow a lot of crucial versatility necessary for complex scenarios. Just another **[reference](https://learn.knack.com/article/zplf1gfk5m-page-rules#what_are_page_rules)** for Page Rules.

I have also written and utilised extensive **[javascript](**url**)** and **[css](https://docs.knack.com/docs/using-css-with-knack)** in all of my apps to enhance the functionalities even further.

A good 'complex form' building experience is definitely a requirement for modern apps, which was only partially satisfied in Knack, for me.

I shall ping you on Discord.

**Comment by FloMiLe** [comment]
I am also excited to read that many of arladmin's suggestions are upcoming, especially using/referring data in a table from other connected tables. In this regard, but also in general, I think it would be helpful to be able to specify relations, i.e. one-to-one, one-to-many and many-to-many. On one hand, this would restrict user input where a only a one-to-one or one-to-many relation is sensible. On the other hand, this would allow for more seamless referring of data.
